src/boxrobots/src/controllers/QPinnerloopcontroller.jl

```diff
@@ -0,0 +1,476 @@
+      return contact_mismatch
+    end
+  else
+    # could have NoMismatch or EarlyContact
+    has_contact_switch = !isequal(contact_state, planned_contact_state)
+
+    if has_contact_switch
+      print("\n\n-----------------------\n\n")
+      print("\n\n----------Early Contact-----------")
+      print("t_plan = ", t_plan)
+      print("\n\n")
+      # we should be in EarlyContact here
+      ~, contact_switch = ContactSwitch(planned_contact_state, contact_state)
+      # determine the contact switch type
+      contact_switch_type = get_contact_switch_type(contact_switch)
+
+      # ensure that contact_state == nxt_planned_contact_state
+      nxt_planned_contact_state = get_next_contact_state(contact_plan)
+
+      # sanity check
+      if !isequal(contact_state, nxt_planned_contact_state)
+        error("had contact switch but next state is different from planned")
+      end
+
+      # sanity check
+      if contact_switch_type != MakingContact
+        error("in EarlyContact case but contact_switch_type isn't MakingContact")
+      end
+
+      # plan stuff will be updated when we switch from EarlyContact to normal contact
+      nxt_contact_mismatch = ContactMismatch(contact_switch.limb_symbol, EarlyContact)
+      return nxt_contact_mismatch
+    else
+      # nothing happened in this case
+      return contact_mismatch
+    end
+  end
+end
+
+function update_contact_mismatch_state!(t_plan::Float64, contact_state::ContactState,
+  contact_plan::ContactPlan, contact_mismatch::ContactMismatch{LateContact})
+  """
+  Check that we are in agreement with the planned contact state. If so then
+  return a NoMismatch. Otherwise throw an error
+  """
+  planned_contact_state = get_contact_state_from_plan(t_plan)
+  if !isequal(contact_state, planned_contact_state)
+    error("In LateContact mode, but contact_states aren't equal")
+  end
+
+  return ContactMismatch(Symbol(), NoMismatch)
+end
+
+function update_contact_mismatch_state!(t_plan::Float64, contact_state::ContactState,
+  contact_plan::ContactPlan, contact_mismatch::ContactMismatch{EarlyContact})
+  """
+  Checks to see if t_plan > nxt_contact_switch_time. If it is then switch to
+  NoMismatch. If there is no next switch then throw an error
+  """
+  if !has_next_contact_switch(contact_plan)
+    error("In state EarlyContact but plan has no more contact switches")
+  end
+
+  nxt_contact_switch_time = get_next_contact_switch_time(contact_plan)
+  if (t_plan >= nxt_contact_switch_time)
+    increment_next_contact_state_idx!(contact_plan)
+    nxt_contact_mismatch = ContactMismatch(Symbol(), NoMismatch)
+  else
+    return contact_mismatch
+  end
+end
+
+function extract_contact_assignments_from_plan(t_current::Float64, controller::QPInnerLoopController)
+  """
+  Gets the contact assigment from plan at knot point times specified by
+  `t_plan, num_time_steps` and `dt`
+
+  # Returns
+  * `contact_assignments`: dict with contact assignment for each limb
+  """
```

Comment:
putting the docstring inside the function is the right thing for Python, but Julia puts the documentation immediately outside the function instead, like:

```
"""
A description of f()
"""
function f() 
  return 0
end
```